[libc++][hardening] Use bounded iterators in std::vector and std::string

davidben opened the pull request:
~~NB: This PR depends on #78876. Ignore the first commit when reviewing, and don't merge it until #78876 is resolved. When/if #78876 lands, I'll clean this up.~~

This partially restores parity with the old, since removed debug build. We now can re-enable a bunch of the disabled tests. Some things of note:

- `bounded_iter`'s converting constructor has never worked. It needs a friend declaration to access the other `bound_iter` instantiation's private fields.

- The old debug iterators also checked that callers did not try to compare iterators from different objects. `bounded_iter` does not currently do this, so I've left those disabled. However, I think we probably should add those. See https://github.com/llvm/llvm-project/issues/78771#issuecomment-1902999181

- The `std::vector` iterators are bounded up to capacity, not size. This makes for a weaker safety check. This is because the STL promises not to invalidate iterators when appending up to the capacity. Since we cannot retroactively update all the iterators on `push_back()`, I've instead sized it to the capacity. This is not as good, but at least will stop the iterator from going off the end of the buffer.

  There was also no test for this, so I've added one in the `std` directory.

- `std::string` has two ambiguities to deal with. First, I opted not to size it against the capacity. https://eel.is/c++draft/string.require#4 says iterators are invalidated on an non-const operation. Second, whether the iterator can reach the NUL terminator. The previous debug tests and the special-case in https://eel.is/c++draft/string.access#2 suggest no. If either of these causes widespread problems, I figure we can revisit.

- `resize_and_overwrite.pass.cpp` assumed `std::string`'s iterator supported `s.begin().base()`, but I see no promise of this in the standard. GCC also doesn't support this. I fixed the test to use `std::to_address`.

- `alignof.compile.pass.cpp`'s pointer isn't enough of a real pointer. (It needs to satisfy `NullablePointer`, `LegacyRandomAccessIterator`, and `LegacyContiguousIterator`.) `__bounded_iter` seems to instantiate enough to notice. I've added a few more bits to satisfy it.

Fixes #78805

davidben committed:
[libc++][hardening] Use bounded iterators in std::vector and std::string

This partially restores parity with the old, since removed debug build.
We now can re-enable a bunch of the disabled tests. Some things of note:

- bounded_iter's converting constructor has never worked. It needs a
  friend declaration to access the other bound_iter instantiation's
  private fields.

- The old debug iterators also checked that callers did not try to
  compare iterators from different objects. bounded_iter does not
  currently do this, so I've left those disabled. However, I think we
  probably should add those. See
  https://github.com/llvm/llvm-project/issues/78771#issuecomment-1902999181

- The std::vector iterators are bounded up to capacity, not size. This
  makes for a weaker safety check. This is because the STL promises not
  to invalidate iterators when appending up to the capacity. Since we
  cannot retroactively update all the iterators on push_back(), I've
  instead sized it to the capacaity. This is not as good, but at least
  will stop the iterator from going off the end of the buffer.

  There was also no test for this, so I've added one in the std
  directory.

- std::string has two ambiguities to deal with. First, I opted not to
  size it against the capacity. https://eel.is/c++draft/string.require#4
  says iterators are invalided on an non-const operation. Second,
  whether the iterator can reach the NUL terminator. The previous debug
  tests and the special-case in https://eel.is/c++draft/string.access#2
  suggest no. If either of these causes widespread problems, I figure we
  can revisit.

- resize_and_overwrite.pass.cpp assumed std::string's iterator supported
  s.begin().base(), but I see no promise of this in the standard. GCC
  also doesn't support this. I fixed the test to use std::to_address.

- alignof.compile.pass.cpp's pointer isn't enough of a real pointer. (It
  needs to satisfy NullablePointer, LegacyRandomAccessIterator, and
  LegacyContiguousIterator.) __bounded_iter seems to instantiate enough
  to notice. I've added a few more bits to satisfy it.

Fixes #78805

davidben commented:
> First, I opted not to size it against the capacity. https://eel.is/c++draft/string.require#4 says iterators are invalided on an non-const operation.

Ran into an interesting case. While this PR doesn't trip it, I looked at adding iterator compatibility checks for https://github.com/llvm/llvm-project/issues/78771#issuecomment-1902999181 and that tripped one of libc++'s existing `std::string` tests:

This test is only valid if we take the position that iterators remain valid when appending up to capacity:

https://github.com/llvm/llvm-project/blob/380ac53dfa05792c6f9fd0a4aba542f8c7e5e17c/libcxx/test/std/strings/basic.string/string.modifiers/string_insert/iter_char.pass.cpp#L31-L32

However, if I'm reading the spec correctly, I don't think that is guaranteed. Yet this test is under "std" and not "libcxx". Thoughts? Should we bound the string iterator to the capacity too?

davidben commented:
Another thing I realized: the trick of using the capacity will slightly impede optimizability of range-for loops. Because that will expand the code into:
```
auto begin = vec.__begin_;
auto end = vec.__end_;
auto end_cap = vec. __end_cap();
auto iter = begin;
while (iter != end) {
  assert(iter != end_cap);
  f(*iter);
  auto(iter != end_cap);
  iter++;
}
```

In order for the compiler to delete those assertions, it needs to know that `begin <= end <= end_cap`, and I think it has no way to learn that a priori. Interestingly, the assertions in `__bounded_iter`'s constructor are enough to do that, but

1. They're disabled by way of `_LIBCPP_ASSERT_INTERNAL`
2. We don't actually want a runtime check on that because this is an invariant that `std::vector` maintains. I.e. no incorrect use of `std::vector` (should of rampant memory unsafety scribbling over it) will break that invariant.
3. Although `_LIBCPP_ASSERT_INTERNAL` turns into `_LIBCPP_ASSUME`, `_LIBCPP_ASSUME` is currently disabled because it breaks optimizations. So we have to either fix that or add `_LIBCPP_ASSUME_FOR_REAL` on those two invariants. :-)

davidben commented:
> NB: This PR depends on #78876. Ignore the first commit when reviewing, and don't merge it until #78876 is resolved. When/if #78876 lands, I'll clean this up.

This is now done. Two remaining open questions, other than generally needing review:

1. Should string iterators be bounded to size or capacity? https://github.com/llvm/llvm-project/pull/78929#issuecomment-1908305723
2. How to help the compiler optimize `std::vector`? https://github.com/llvm/llvm-project/pull/78929#issuecomment-1936582711.

For (1), I'm leaning towards leaving it as-is and seeing it it causes problems. If I'm reading the spec right, we're not obligated to keep that working.

For (2), I propose we land https://github.com/llvm/llvm-project/pull/91801 and then, in `vector::begin()` and `vector::end()`, add `_LIBCPP_ASSUME(size <= capacity)`. I believe that'll be enough for the compiler to optimize it, though I still need to confirm this.

@var-const, WDYT?

ldionne commented on libcxx/test/libcxx/containers/sequences/vector/debug.iterator.add.pass.cpp:
```diff
@@ -10,36 +10,48 @@
 
```
These tests should be renamed to `assert.iterator.whatever`

ldionne reviewed (changes requested):
I have a few comments but I think the approach is fine. If you are unable to update this before next Tuesday, please let me know and we will apply the review comments. It would be nice to get this in LLVM 19.

davidben commented:
> If you are unable to update this before next Tuesday, please let me know and we will apply the review comments.

Alas the IETF meeting is next week so I probably won't have time for a bit. If you've got the cycles, go for it! Thanks!

ldionne commented on libcxx/include/vector:
```diff
@@ -401,9 +402,16 @@ public:
   typedef typename __alloc_traits::const_pointer const_pointer;
-  // TODO: Implement iterator bounds checking without requiring the global database.
+#ifdef _LIBCPP_ABI_BOUNDED_ITERATORS
+  // The pointer must be passed through __wrap_iter because
+  // __alloc_traits::pointer may not be detected as a continguous iterator on
+  // its own.
```
```suggestion
  // The pointer must be passed through __wrap_iter because
  // __alloc_traits::pointer may not be detected as a continguous iterator on
  // its own before C++20.
```

ldionne commented on libcxx/include/string:
```diff
@@ -786,9 +787,16 @@ public:
-  // TODO: Implement iterator bounds checking without requiring the global database.
+#ifdef _LIBCPP_ABI_BOUNDED_ITERATORS
+  // The pointer must be passed through __wrap_iter because
+  // __alloc_traits::pointer may not be detected as a continguous iterator on
+  // its own.
```
```suggestion
  // __alloc_traits::pointer may not be detected as a continguous iterator on
  // its own before C++20.
```

var-const committed:
Address feedback plus minor tweaks:

- create dedicated ABI macros
  `_LIBCPP_ABI_BOUNDED_ITERATORS_IN_STRING` and
  `_LIBCPP_ABI_BOUNDED_ITERATORS_IN_VECTOR` (and the associated Lit
  features, etc.);
- add a release note and update the hardening documentation;
- add a comment in `<string>` to mention that its bounded iterators use
  the size rather than the capacity;
- reword and slightly expand the comment in `<vector>` explaining the
  use of the capacity rather than the size;
- tweak the comment that justifies the use of `__wrap_iter`;
- in `string`, move the conditional logic from outside to inside the
  functions' scope;
- move the repeated `fill_to_capacity` function to a separate helper
  header and slightly reword the comment;
- make `push_back.invalidation.pass.cpp` a death test and add a comment
  on some limitations of what a bounded iterator can currently detect;
- fix another `small_pointer` helper class to fix compilation.

var-const committed:
Fix string formatting

var-const commented:
> > If you are unable to update this before next Tuesday, please let me know and we will apply the review comments.
> 
> Alas the IETF meeting is next week so I probably won't have time for a bit. If you've got the cycles, go for it! Thanks!

@davidben Awesome, thanks for the heads-up! I have applied most of the review feedback (plus some of the feedback I had). If you have the time, please take a look and let us know if there's anything you'd like to change/push back on; otherwise, I plan to merge by the Tuesday branch cutoff.

davidben commented:
> If you have the time, please take a look and let us know if there's anything you'd like to change/push back on; otherwise, I plan to merge by the Tuesday branch cutoff.

Feedback all looks reasonable to me. Thanks!

var-const committed:
Merge branch 'main' into bounded-iter-vec-str

var-const commented on libcxx/include/string:
```diff
@@ -786,9 +787,16 @@ public:
                 "original allocator");
 
-  // TODO: Implement iterator bounds checking without requiring the global database.
+#ifdef _LIBCPP_ABI_BOUNDED_ITERATORS
```
Done (leaving the comment open as a reminder that we need to do some further cleanup).

var-const committed:
Add a forgotten test file

var-const committed:
More comments and docs

var-const committed:
Rename assertion test files

var-const committed:
Tweak the release note to make it clear the ABI macros are intended only
for vendors.

var-const committed:
Minor fixes

var-const commented:
(I think everything I planned to do is uploaded now, I plan to merge before tomorrow's cutoff time if no objections)

var-const reviewed (approved):
@davidben Thank you so much for working on this! (And my sincere apologies for the very slow review -- I wanted to get to this for a while but unfortunately had my plate very full recently) LGTM.

My biggest concern starting this review was whether accepting this lightweight approach would prevent us from adding heavier iterators in the future that would retain a connection with the container and thus could also catch temporal safety issues, such as invalidation. That would bring the current incarnation of hardening on par with the historical debug mode. However, since we decided to create separate ABI macros for these new iterators, this is no longer an issue. While it's a tradeoff (having several independent ABI configuration options creates many possible combinations, increasing complexity), I believe it's probably an inevitable outcome -- for one, I don't think it's reasonable to force anyone using the current `bounded-iterators` macro to start using bounded iterators in their vectors and strings, especially with no targeted opt-out possible, but also I think there might be legitimate contexts where someone might want bounded iterators in some, but not all, containers. The ability to choose (in the future, potentially) between more and less lightweight iterator hardening strategies follows quite nicely from this, kind of mirroring the distinction between the `fast` and `extensive` modes.

The idea to bound the vector iterator by capacity has grown on me. It is slightly unfortunate that there is a discrepancy between how checking is performed for strings versus vectors, but it seems like a necessary evil -- the only way to reconsolidate would be to make string checks laxer. Checking just the "physical" bounds prevents the worst vulnerabilities, so I believe we're still getting the most valuable part of the checks; I'd also note that IIUC, this is how Asan checks work for contiguous containers that are not specially annotated for the sanitizer.

I think that even without covering iterator invalidation (which, being a temporal safety issue, is much harder to solve in general anyway), this patch provides a noticeable improvement in the amount of checks we provide with little added code complexity, and the fact that the bound iterators are comparatively lightweight can be an advantage for some contexts.
